src/Gui/OverlayManager.cpp

```diff
@@ -0,0 +1,2061 @@
+                    OverlayParams::getDockOverlayRevealDelay()));
+        refresh();
+    }
+
+    void onFocusChanged(QWidget *, QWidget *) {
+        refresh();
+    }
+
+    void setupTitleBar(QDockWidget *dock)
+    {
+        if(!dock->titleBarWidget())
+            dock->setTitleBarWidget(createTitleBar(dock));
+    }
+
+    QWidget *createTitleBar(QWidget *parent)
+    {
+        OverlayTitleBar *widget = new OverlayTitleBar(parent);
+        widget->setObjectName(QStringLiteral("OverlayTitle"));
+        QList<QAction*> actions;
+        if (auto tabWidget = qobject_cast<OverlayTabWidget*>(parent))
+            actions = tabWidget->actions();
+        else
+            actions = _actions;
+        widget->setTitleItem(OverlayTabWidget::prepareTitleWidget(widget, actions));
+        return widget;
+    }
+
+    void onAction(QAction *action) {
+        if(action == &_actOverlay) {
+            OverlayManager::instance()->setOverlayMode(OverlayManager::OverlayMode::ToggleActive);
+        } else if(action == &_actFloat || action == &_actClose) {
+            for(auto w=qApp->widgetAt(QCursor::pos());w;w=w->parentWidget()) {
+                auto dock = qobject_cast<QDockWidget*>(w);
+                if(!dock)
+                    continue;
+                setFocusView();
+                if(action == &_actClose) {
+                    dock->toggleViewAction()->activate(QAction::Trigger);
+                } else {
+                    auto it = _overlayMap.find(dock);
+                    if(it != _overlayMap.end()) {
+                        it->second->tabWidget->removeWidget(dock);
+                        getMainWindow()->addDockWidget(it->second->dockArea, dock);
+                        _overlayMap.erase(it);
+                        dock->show();
+                        dock->setFloating(true);
+                        refresh();
+                    } else 
+                        dock->setFloating(!dock->isFloating());
+                }
+                return;
+            }
+        } else {
+            auto tabWidget = qobject_cast<OverlayTabWidget*>(action->parent());
+            if(tabWidget)
+                tabWidget->onAction(action);
+        }
+    }
+
+    void retranslate()
+    {
+        _actOverlay.setToolTip(QObject::tr("Toggle overlay"));
+        _actFloat.setToolTip(QObject::tr("Toggle floating window"));
+        _actClose.setToolTip(QObject::tr("Close dock window"));
+    }
+
+    void floatDockWidget(QDockWidget *dock)
+    {
+        setFocusView();
+        auto it = _overlayMap.find(dock);
+        if (it != _overlayMap.end()) {
+            it->second->tabWidget->removeWidget(dock);
+            _overlayMap.erase(it);
+        }
+        dock->setFloating(true);
+        dock->show();
+    }
+
+    // Warning, the caller may be deleted during the call. So do not pass
+    // parameter using reference, pass by value instead.
+    void dragDockWidget(QPoint pos,

```

Comment:
This function is _400_ lines long... it's pretty hard to argue that could possibly not violate the Single-Responsibility Principle.